Added orchestrator-infra subchart

ElaiShalevRH opened the pull request:
This PR will introduce the orchestrator-infra helm chart, that will be included as a dependency for RHDH chart for the RHDHxOrchestrator flavor of RHDH. This PR will introduce changes to the Chart Values: upon enabling orchestrator, the RHDH chart will install the orchestrator-infra chart before the backstage chart.  

The orchestrator-infra chart will incluse template files for serverless operator (knative eventing, serving) and openshift-serverless operator (sonataflow) that are pre-requisite infrastructure for Orchestrator to work.  

Tests and documentation are TBA. 

Relates to this epic: https://issues.redhat.com/browse/RHIDP-6159 

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the `values.yaml` and added to the README.md. The [pre-commit](https://pre-commit.com/) utility can be used to generate the necessary content. Use `pre-commit run -a` to apply changes.
- [x] JSON Schema template updated and re-generated the raw schema via `pre-commit` hook.
- [x] List tests pass for Chart using the [Chart Testing](https://github.com/helm/chart-testing) tool and the `ct lint` command.

masayag commented on charts/orchestrator-infra/Chart.yaml:
```diff
@@ -0,0 +1,18 @@
+apiVersion: v2
+name: orchestrator-infra
+description: >
+  Helm chart to deploy the Orchestrator solution's required infrastructure suite on OpenShift, including Janus IDP backstage, SonataFlow Operator, OpenShift Serverless Operator, Knative Eventing and Knative Serving.
```
s/SonataFlow/OpenShift Serverless Logic

masayag replied:
also `Janus IDP backstage` should be removed from the list.

ElaiShalevRH commented:
@masayag Updated the PR with all the comments

masayag commented on charts/backstage/values.yaml:
```diff
@@ -337,3 +337,10 @@ test:
+
+orchestrator:
+  enabled: true
+  sonataFlowOperator:
```
s/sonataFlowOperator/serverlessLogic
or go full name for all operators

masayag commented on charts/orchestrator-infra/README.md:
```diff
@@ -0,0 +1,25 @@
+
+![Version: 0.0.1](https://img.shields.io/badge/Version-0.0.1-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square)
+
+Helm chart to deploy the Orchestrator solution's required infrastructure suite on OpenShift, including Janus IDP backstage, SonataFlow Operator, OpenShift Serverless Operator, Knative Eventing and Knative Serving.
```
s/SonataFlow/OPenShift Serverless Logic

masayag commented on charts/orchestrator-infra/values.yaml:
```diff
@@ -0,0 +1,18 @@
+    installPlanApproval: Manual # whether the update should be installed automatically
+    name: logic-operator-rhel8 # name of the operator package
+    sourceName: redhat-operators # name of the catalog source
+    startingCSV: logic-operator-rhel8.v1.33.0 # The initial version of the operator
```
not v1.35.0 ?

masayag commented on charts/orchestrator-infra/values.yaml:
```diff
@@ -0,0 +1,18 @@
+    channel: stable # channel of an operator package to subscribe to
+    installPlanApproval: Manual # whether the update should be installed automatically
+    name: serverless-operator # name of the operator package
+    sourceName: redhat-operators # name of the catalog source
```
no need for startingCSV for this operator?

ElaiShalevRH replied:
didn't have that field in the helm based chart

masayag commented:
@rm3l can you take a look pls?

rm3l commented:
> @rm3l can you take a look pls?

Sure - I'll take a look soon.

/cc

rm3l commented:
FYI

/cc @durandom

ElaiShalevRH commented on charts/backstage/Chart.yaml:
```diff
@@ -26,6 +26,10 @@ dependencies:
     repository: https://backstage.github.io/charts
     version: "2.4.0"
     alias: upstream
+  - name: orchestrator-infra
```
Are these two mutually exclusive? The Infra chart here is logically separate form the RHDH chart, not using and common template or values. If it were to be ship it somewhere else the only thing we need to change is it's location and the repository field, right?
Regardless, can we ship and maintain it separately but still have it reside in the RHDH chart repository? as it won't have any implications for any other product

rm3l replied:
> Are these two mutually exclusive? The Infra chart here is logically separate form the RHDH chart, not using and common template or values. If it were to be ship it somewhere else the only thing we need to change is it's location and the repository field, right?

The code separation makes sense.
Maybe I'm missing something, but while testing this PR, I was not able to run `helm upgrade --install` twice with `orchestrator.enabled` set to `true`. The first install worked (by correctly installing the required operators), but when I tried to upgrade the Helm Release, it failed.
So I am concerned that by having this hard dependency, users may not be able to upgrade their RHDH Orchestrator instances once created.
Also, I think removing the dependency also clarifies the responsibilities: the cluster admin would need to deploy the orchestrator-infra chart as a dependency, and regular users would be able to install their RHDH Orchestrator instances.
Also, IIRC, there was another point that we discussed. By shipping the orchestrator-infra chart separately, we could reuse it in the RHDH Operator when working on the Operator-based Orchestrator flavor.

> If it were to be ship it somewhere else the only thing we need to change is it's location and the repository field, right?
> Regardless, can we ship and maintain it separately but still have it reside in the RHDH chart repository? as it won't have any implications for any other product

And yes, I think it can still reside in this repo as you did, and we can build and ship it separately. See our current downstream pipeline that ships the RHDH Chart: it works on the `charts/backstage` folder, but I guess we can do something similar with the new `charts/orchestrator-infra` folder: https://gitlab.cee.redhat.com/rhidp/rhdh/-/blob/rhdh-1-rhel-9/build/helm/prepare.sh
IMO, it could be shipped in the same official [OpenShift Helm Charts repo](https://charts.openshift.io/) at https://github.com/openshift-helm-charts/charts/tree/main/charts/redhat/redhat

Thoughts?

ElaiShalevRH replied:
> by having this hard dependency, users may not be able to upgrade their RHDH Orchestrator instances once created.

This is a valid concern which I missed before. I agree it makes sense to remove that dependency and keep it as a pre-requisite for admins to take care of in advance. 

So in practice,
1. I'll change this PR so we keep the orchestrator-infra chart here under charts/, but have it shipped seperately in https://github.com/openshift-helm-charts/charts/tree/main/charts/redhat/redhat. 
2. The logic to publish this chart will be added to https://gitlab.cee.redhat.com/rhidp/rhdh/-/blob/rhdh-1-rhel-9/build/helm/prepare.sh 
3. We remove the orchestrator-infra dependency from the RHDH chart, and document the option for admin installations.  

Does this sound correct?

rm3l replied:
Yup, sounds good. Thanks @ElaiShalevRH !

rm3l replied:
@masayag Does that make sense to you?

masayag replied:
yes, each chart should be installed on its own.
The infra chart by the cluster admin with an elevated set of permissions, and the rhdh-chart that can enable the orchestrator with permissions in namespace scope.
---
as part of the verification, we should install the infra chart once, and install the rhdh chart in few namespaces.

masayag commented:
@ElaiShalevRH pls add also notes.txt file  - https://helm.sh/docs/chart_template_guide/notes_files/
so the user that install the chart can verify it was properly installed.

ElaiShalevRH commented:
Hey @rm3l @masayag 
Updated the Infra chart, now with no relation to the RHDH chart. 

Should I open a MR for updating https://gitlab.cee.redhat.com/rhidp/rhdh/-/blob/rhdh-1-rhel-9/build/helm/prepare.sh to include the orchestrator-chart build?

Also added NOTES.txt

rm3l commented on charts/orchestrator-infra/templates/openshift-serverless-logic.yaml:
How about splitting this file into different template files (one per resource)? I think it would help.

rm3l commented on charts/orchestrator-infra/templates/openshift-serverless.yaml:
Same point on splitting this file into several templates.

rm3l commented on charts/orchestrator-infra/values.yaml:
What are the expectations if a user sets both `serverlessLogicOperator.enabled` and `serverlessOperator.enabled` to `false`? Do we expect Orchestrator users to potentially deploy one of the Operators without the other one?

ElaiShalevRH replied:
I think a user could have an existing serverlessLogicOperator or serverlessOperator without the other, so the chart should be able to only install one of them.

rm3l reviewed (commented):
Would it be possible to also add a [`test` template](https://helm.sh/docs/topics/chart_tests/)? I also think that would be required for shipping the Chart into the OpenShift Charts repo.

rm3l commented:
> Should I open a MR for updating [gitlab.cee.redhat.com/rhidp/rhdh/-/blob/rhdh-1-rhel-9/build/helm/prepare.sh](https://gitlab.cee.redhat.com/rhidp/rhdh/-/blob/rhdh-1-rhel-9/build/helm/prepare.sh) to include the orchestrator-chart build?
> 

@ElaiShalevRH That task was captured in https://issues.redhat.com/browse/RHIDP-6169. We (RHDH team) might be able to handle this, so you can focus on getting your PRs merged on GH.
I'll check with the team next week if anyone can help with this. I'll get back to you.

rm3l reviewed (commented):
@ElaiShalevRH Could you please rebase your PR branch? It should hopefully fix the CI errors: https://github.com/redhat-developer/rhdh-chart/actions/runs/13836562023/job/38726590327?pr=108#step:11:30
Thanks.

ElaiShalevRH commented on charts/orchestrator-infra/templates/openshift-serverless-logic.yaml:
```diff
@@ -0,0 +1,32 @@
+  name: {{ .Values.serverlessLogicOperator.subscription.name }}
+  namespace: {{ .Values.serverlessLogicOperator.subscription.namespace }}
+spec:
+  channel: {{ .Values.serverlessLogicOperator.subscription.channel }}
```
Hey, this makes perfect sense. I'll implement what you suggested

ElaiShalevRH commented:
> > Should I open a MR for updating [gitlab.cee.redhat.com/rhidp/rhdh/-/blob/rhdh-1-rhel-9/build/helm/prepare.sh](https://gitlab.cee.redhat.com/rhidp/rhdh/-/blob/rhdh-1-rhel-9/build/helm/prepare.sh) to include the orchestrator-chart build?
> 
> @ElaiShalevRH That task was captured in https://issues.redhat.com/browse/RHIDP-6169. We (RHDH team) might be able to handle this, so you can focus on getting your PRs merged on GH. I'll check with the team next week if anyone can help with this. I'll get back to you.

Thanks, that could be helpful
If you don't seem to have time for it I could also take that task, of course

ElaiShalevRH commented:
Hey @rm3l, Thanks for the review :) 
I've implemented all the changes you've suggested. The chart is ready to be reviewed again. 
I'm also working on a test, it looks like its clashing with the SonadCloud analysis, so I've deleted it for now but will keep working on it. 

Please let me know if there is more to be done that is required for shipping the Chart. 
@masayag

rm3l commented on charts/orchestrator-infra/values.yaml:
```diff
@@ -0,0 +1,24 @@
+  subscription:
+    namespace: openshift-serverless-logic  #  namespace where the operator should be deployed
+    channel: alpha  #  channel of an operator package to subscribe to
+    installPlanApproval: Manual  #  whether the update should be installed automatically
```
```suggestion
    installPlanApproval: Automatic  #  whether the update should be installed automatically
```

Shouldn't this be Automatic by default?

ElaiShalevRH replied:
Having the installPlan be Automatic will cause it to automatically install updates as well. We want to avoid scenarios where the OpenshiftServerlessLogic operator version gets upgraded, but the newer version isn't supported by other orchestrator functionality.

rm3l replied:
Ok, that makes sense. So there would be an extra step for cluster admins (approving the Install Plan) to not forget to document.

rm3l commented on charts/orchestrator-infra/values.yaml:
```diff
@@ -0,0 +1,24 @@
+  subscription:
+    namespace: openshift-serverless  # namespace where the operator should be deployed
+    channel: stable  #  channel of an operator package to subscribe to
+    installPlanApproval: Manual  #  whether the update should be installed automatically
```
```suggestion
    installPlanApproval: Automatic  #  whether the update should be installed automatically
```

Same point on making this Automatic by default..

rm3l commented on charts/orchestrator-infra/templates/NOTES.txt:
```diff
@@ -0,0 +1,38 @@
+====================================================================
+{{/* Empty line */}}
+
+To manually approve the InstalPlans:
```
Can this and the commands below be shown only if `.Values.serverless{,Logic}Operator.subscription.installPlanApproval` are `Manual`?

rm3l commented on charts/orchestrator-infra/templates/serverless/subscription.yaml:
```diff
@@ -0,0 +1,10 @@
+  name: {{ .Values.serverlessOperator.subscription.name }}
+  namespace: {{ .Values.serverlessOperator.subscription.namespace }}
+spec:
+  {{- toYaml .Values.serverlessOperator.subscription | nindent 2 }}
```
I suggested moving the subscription spec under dedicated `serverlessOperator.subscription.spec` and `serverlessLogicOperator.subscription.spec` fields in the values file. Otherwise, users might get warnings because some fields are not supposed to be part of the Subscription spec:
```
Release "my-orchestrator-infra-1" does not exist. Installing it now.
W0317 10:06:28.037562 1919983 warnings.go:70] unknown field "spec.namespace"
W0317 10:06:28.037564 1919983 warnings.go:70] unknown field "spec.namespace"
NAME: my-orchestrator-infra-1
[...]
```
WDYT?

ElaiShalevRH replied:
Ah, got it. Yeah I see what you mean, I'll sort that out

rm3l reviewed (commented):
@ElaiShalevRH Thanks for the changes. A few comments below ;)

rm3l commented on charts/orchestrator-infra/Chart.yaml:
```diff
@@ -0,0 +1,15 @@
+charts.openshift.io/name: Orchestrator-Infra
+charts.openshift.io/provider: Red Hat Developer Hub Team
+charts.openshift.io/supportURL: https://github.com/redhat-developer/rhdh-chart/issues
```
```suggestion
charts.openshift.io/supportURL: https://issues.redhat.com/browse/RHIDP
```

I also realized that we should be pointing to JIRA instead since we purposely disabled GH issues in this repo. It is being fixed in the `backstage` charts as well (https://github.com/redhat-developer/rhdh-chart/pull/114)

ElaiShalevRH commented on charts/orchestrator-infra/templates/tests/infra-test.yaml:
```diff
@@ -0,0 +1,83 @@
+  annotations:
+    helm.sh/hook: test
+spec:
+  serviceAccountName: test-service-account
```
@rm3l Any help in getting this sorted? I'm not sure what it expects, there is a ClusterRole and a ClusterRoleBinding, and the resource get deleted after the test

rm3l replied:
Indeed, this is weird.. Not sure why.. Is it because of the ClusterRole and ClusterRoleBinding, which provide wide access to some of the resources?
Can you include a Role and RoleBinding in the Release namespace, so that SonarCloud sees that the Service account permissions are somehow restricted. Hopefully that would help 🤞🏿

rm3l commented on charts/orchestrator-infra/templates/tests/infra-test.yaml:
```diff
@@ -0,0 +1,83 @@
+        - |
+          set -e
+          
+          kubectl get subscription -n {{ .Values.serverlessOperator.subscription.spec.namespace }} || exit 1
```
I guess this should be executed only if `.Values.serverlessOperator.enabled` is `true`? Same point for the subsequent lines..

rm3l replied:
And would it make sense to get the subscription that the Helm Chart created?
```suggestion
          kubectl get subscription "{{ .Values.serverlessOperator.subscription.spec.name }}" -n {{ .Values.serverlessOperator.subscription.spec.namespace }} || exit 1
```

rm3l commented on charts/orchestrator-infra/templates/tests/infra-test.yaml:
```diff
@@ -0,0 +1,83 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: "testing-pod"
```
I understand that the user may not be able to deploy this infra Chart twice, but a general rule of thumb is that namespace-scoped resources should contain the Helm Release name in their name.

rm3l reviewed (commented):
Almost there ;-)

ElaiShalevRH commented:
@rm3l Hey, added some more changes
The "Bind this Service Account to RBAC or disable "automountServiceAccountToken" error persists... I've changes all ClusterRoles[binding] to namespaced Role, Rolebindins, it still complains...

rm3l commented:
> @rm3l Hey, added some more changes The "Bind this Service Account to RBAC or disable "automountServiceAccountToken" error persists... I've changes all ClusterRoles[binding] to namespaced Role, Rolebindins, it still complains...

Okay, sounds like a false positive to me. I've dismissed it for now. Let's see how it goes now.

rm3l reviewed (commented):
LGTM. Thanks for your patience, @ElaiShalevRH !! If you don't have anything else to add to this PR (noticed the "WIP" in the PR title), I am fine with merging this first version of the orchestrator infra chart.

rm3l reviewed (approved):
/lgtm
